feature_engine/timeseries/forecasting/window_features.py

```diff
@@ -0,0 +1,333 @@
+from feature_engine.variable_manipulation import (
+    _check_input_parameter_variables,
+    _find_or_check_numerical_variables,
+    _find_all_variables,
+)
+
+
+@Substitution(
+    variables=_variables_numerical_docstring,
+    missing_values=_missing_values_docstring,
+    drop_original=_drop_original_docstring,
+    feature_names_in_=_feature_names_in_docstring,
+    n_features_in_=_n_features_in_docstring,
+    fit=_fit_not_learn_docstring,
+    fit_transform=_fit_transform_docstring,
+)
+class WindowFeatures(BaseEstimator, TransformerMixin):
+    """
+        LagFeatures adds lag features to the dataframe. A lag feature is a feature with
+        information about a prior time step.
+
+        LagFeatures has the same functionality as pandas `shift()` with the exception that
+        only one of `periods` or `freq` can be indicated at a time. LagFeatures builds on
+        top of pandas `shift()` in that multiple lags can be created at the same time and
+        the features with names will be concatenated to the original dataframe.
+
+        To be compatible with LagFeatures, the dataframe's index must have unique values
+        and no NaN.
+
+        LagFeatures works only with numerical variables. You can pass a list of variables
+        to lag. Alternatively, LagFeatures will automatically select and lag all numerical
+        variables found in the training set.
+
+        More details in the :ref:`User Guide <lag_features>`.
+
+        Parameters
+        ----------
+        {variables}
+
+        periods: int, list of ints, default=1
+            Number of periods to shift. Can be a positive integer or list of positive
+            integers. If list, features will be created for each one of the periods in the
+            list. If the parameter `freq` is specified, `periods` will be ignored.
+
+        freq: str, list of str, default=None
+            Offset to use from the tseries module or time rule. See parameter `freq` in
+            pandas `shift()`. It is the same functionality. If freq is a list, lag features
+            will be created for each one of the frequency values in the list. If freq is not
+            None, then this parameter overrides the parameter `periods`.
+
+        sort_index: bool, default=True
+            Whether to order the index of the dataframe before creating the lag features.
+
+        {missing_values}
+
+        {drop_original}
+
+        Attributes
+        ----------
+        variables_:
+            The group of variables that will be lagged.
+
+        {feature_names_in_}
+
+        {n_features_in_}
+
+        Methods
+        -------
+        {fit}
+
+        transform:
+            Add lag features.
+
+        {fit_transform}
+
+        See Also
+        --------
+        pandas.shift
+        """
+
+    def __init__(
+            self,
+            variables: List[str] = None,
+            window: Union[str, int] = 1,
+            function: Callable = np.mean,
```

Comment:
I have two questions for you guys @Morgan-Sell  and @KishManani

First, in other transformers I call this parameter "operations" instead of "functions". I'd like to unify the way we call parameters that do similar things throughout the package. So my question is: which is a better name for this parameter? "operations", "functions", something else, or it does not really matter? If it is not important, I would prefer "operations" so I don't have to deprecate in the other transformer.

And second, I see that @Morgan-Sell is passing numpy functions here. But pandas agg() takes strings. So, should this parameter take strings, numpy functions, or we don't care because in any case, pandas takes care of this check? and we just allow everything?

Quick note: whenever functionality is already implemented in packages that have great developer support, like pandas, sklearn and scipy, I tend to leave "the work" to them (that is, their methods and classes), because we are a tiny community in comparison, so maintaining for us is harder.

Thank you!

Reply:
I think we need to consider what kind of transformations we want to accept. `agg` is orders of magnitude faster for  operations like sum and mean relative to `apply` (like ms for `agg` vs seconds for `apply`). Also the `agg` API is a bit more flexible in allowing you to specify multiple different transformations for different variables using a dictionary (see second highest comment (here)[https://stackoverflow.com/questions/21828398/what-is-the-difference-between-pandas-agg-and-apply-function]).

Reply:
In https://github.com/Morgan-Sell/feature_engine/pull/9 I modified the functionality of the class so that it works with `agg` instead of `apply`. 

Agg can take strings representing each one of the rolling() allowed aggregation functions. That is convenient.

What the current functionality is not allowing, is for the user to pass a bespoke function. Which might be upsetting.

We have one of 2 options:

1. release like this and wait for user feedback
2. try to think how we can allow the use of bespoke functions

For 2, it would be helpful if @KishManani has an example of what this could be?

Reply:
Bespoke functions to pass to `agg` is anything that aggregates the time series into a scalar value (e.g., a quantile, the kurtosis, the entropy).

Reply:
Also, allowing a custom function allows you to do weighted averages too:

```
weights = np.array([0.1, 0.2, 0.3, 0.4])
df.rolling(window=4).agg(lambda x: np.sum(x * weights))
```